### ISSUE TITLE
Impossible intersection is supposedly never but any can still be assigned to it

### ISSUE THREAD

**Issue by benjaminjkraft** [microsoft/TypeScript#53027]
# Bug Report

### 🔎 Search Terms

intersection assign never

### 🕗 Version & Regression Information

This is the behavior in every version I tried (including nightly), and I reviewed the FAQ for relevant entries. But really this is a loose end left by #36696; prior to that the impossible intersection (`AB` below) was not `never` at all.

### ⏯ Playground Link

[link](https://www.typescriptlang.org/play?ts=4.9.5#code/C4TwDgpgBAglC8UDeUDWBLAdgEwFxQHIBDAgGigDMB7K-AZ2ACcsBzKAXwG4AoUSKAEIJkaLHkIAjMpRr5MAVwC2EiIw49e4aDCGI4AMkGcoUAPSmomCADdV5OvLBgqdCNgA2Ibt3NQAAsB0ALQQAB6QAMbAIYyMVIz4ACpahESYIARQ6HSWVMBQRHR06CyYRBLu0MBUUHzQBFa2jAQAdNwRVJgMlgCMcjaqwkjsBTlpXr4S8vnAABbZWTlUqO2d3ZgATPg6QyOFBelAA)

### 💻 Code

(modified from the example in #36696)

```ts
type A = { kind: 'a', foo: string };
type B = { kind: 'b', foo: number };

type AB = A & B;  // never, supposedly

// @ts-expect-error: Type 'any' is not assignable to type 'never'.
const n1: never = {} as any
// but this is ok
const n2: AB = {} as any
```

### 🙁 Actual behavior

`any` is assignable to `AB`, despite the fact that `any` is not assignable to `never` and `AB` is supposedly `never` (per both #36696 and hovering it)

### 🙂 Expected behavior

`any` should not be assignable to `AB`

**Comment by RyanCavanaugh** [comment]
Is this behavior observable in a way that matters?

**Comment by benjaminjkraft** [comment]
Two problems I know of that seem to be caused by this.

First, it means assigning to the quasi-never type `AB` gives errors on each field rather than on the object as a whole, which makes those errors harder to understand IMO:
```ts
const n3: AB = {
    // @ts-expect-error: Type 'number' is not assignable to type 'never'.
    a: 3,
}
```

Second, it means that `Omit<AB, K>` gives an incorrect type (even when using the conditional type trick such that it does work with literal `never`:
```ts
type DistributedOmit<T, K extends keyof any> = T extends any ? Omit<T, K> : never
// @ts-expect-error: Type '{}' is not assignable to type 'never'.
const n4: DistributedOmit<never, "bogus"> = {}
// allowed
const n5: DistributedOmit<AB, "bogus"> = {}
```

(To be clear, these are all simplified from something I ran into in real code.)

Full playground [here](https://www.typescriptlang.org/play?ts=4.9.5#code/C4TwDgpgBAglC8UDeUDWBLAdgEwFxQHIBDAgGigDMB7K-AZ2ACcsBzKAXwG4AoUSKAEIJkaLHkIAjMpRr5MAVwC2EiIw49e4aDCGI4AMkGcoUAPSmomCADdV5OvLBgqdCNgA2Ibt3NQAAsB0ALQQAB6QAMbAIYyMVIz4ACpahESYIARQ6HSWVMBQRHR06CyYRBLu0MBUUHzQBFa2jAQAdNwRVJgMlgCMcjaqwkjsBTlpXr4S8vnAABbZWTlUqO2d3ZgATPg6QyOFBenevgDu8xGzUIoQaTmzqtDH0CwQ+apxjEuYUNfnUGBxkEYoCyXWA12wUCoFEhXzmD1mVEqkIkACsIFFVqDLABmba6ZDcEwmXwBYJhSLRN7xJIpBpKFTNRa5fKFYqlcpI6q1WmNVStQlEoj4bGkbjsTT8AAi2SY6CmYOwAHlFOhgAAeRLkADS31CYJwOVQEBAUIOIAAfMJErr9dgxukoAB+KDK1Ua7WW-pNHwWIjuOg1VwQRQ5LkRIjyVxQaUMZjytyumY1TB5KDHeKoKDMFizYC4H3+QIhcLoymxalQZL8AjDTILFMsoolMoVKo1OqEXnNNodLGYAAs+BjsvjSpV6q75AARBIqCxI1PLYhhgW-e4qI9sJj1gBWIcyuPTBPjtU6aez+d0Re7IA).

**Comment by RyanCavanaugh** [comment]
The presented issue was that you don't want `any` to be assignable to `AB`, though, which seems totally unrelated.

**Comment by benjaminjkraft** [comment]
Sure, I guess that was intended as the simplest example of "this type is supposedly never but is not in fact equivalent to never". If it's more useful, I'm happy to make a new issue framed around those examples.

**Comment by RyanCavanaugh** [comment]
Yeah, the exact manifestation / what you care about matters. These are all orthogonally addressable, in theory.

The PR notes that this is a consequence of the reduction being deferred:

> In https://github.com/microsoft/TypeScript/pull/31838 we reduce empty intersections to never immediately upon construction. However, in this PR we have to defer the determination because it requires us to resolve the types of properties in the constituent object types. Specifically, when two or more constituent object types have properties with the same name, we need to resolve the types of those properties in order to determine if they're disjoint discriminants, but doing so during intersection construction can easily cause circularities. So, instead of reducing intersections upon construction, we reduce them immediately before accessing members, relating them to other types, or converting them to their string representations in diagnostics and quick info.

I'm not sure if there are any cures on the table which are better than the disease here. Basically, not all of these can be true at once:

 * Nothing (except `never`) is assignable to `never`
 * `S` is assignable to `T & U` if `S` is assignable to `T` and `S` is assignable to `U`
 * `any` is assignable to anything that isn't `never`
 * Just mentioning the type `A & B` won't introduce new circularity errors into your code

so we have to park the inconsistency somewhere. It's extremely rare to see a zero-order intersection that reduces to `never` in practice, so choosing that scenario as the one to be the most consistent at the expense of violating other invariants seems like a bad choice. I'd be curious to know more about how you ended up passing around an unreduced intersection that was actually `never`; these typically only happen in higher-order scenarios where they get reduced out of the final union before they can land in usercode and cause problems.

IMO probably the most convincing example that something is wrong is something like this:
```ts
// x is effectively { [s: string]: never }
type X = Omit<A & B, "nada">;
const x: X = { }; // Allowed initialization, maaaaybe ok per definition of X?
const s: string = x.anything; // Allowed property access, sad
console.log(s.toLowerCase()); // Crash, sad.
```
but the initialization of `X` only succeeds if the initializer is exactly `{ }`; anything else would fail, same as here
```ts
// @ts-expect-error: Type '{}' is not assignable to type 'never'.
const n4: DistributedOmit<never, "bogus"> = {}
```
which is sort of a niche problem.
